pkg/gardenlet/controller/federatedseed/network-policy/controller.go

```diff
@@ -0,0 +1,202 @@
+	var (
+		endpointsInformer = seedDefaultNamespaceKubeInformer.Core().V1().Endpoints()
+		endpointsLister   = endpointsInformer.Lister()
+
+		networkPoliciesInformer = seedKubeInformerFactory.Networking().V1().NetworkPolicies()
+		networkPoliciesLister   = networkPoliciesInformer.Lister()
+
+		namespaceInformer = seedKubeInformerFactory.Core().V1().Namespaces()
+		namespaceLister   = namespaceInformer.Lister()
+		namespaceQueue    = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "namespace")
+	)
+
+	controller := &Controller{
+		log:                 seedLogger,
+		namespaceReconciler: newNamespaceReconciler(ctx, seedLogger, seedClient.Client(), endpointsLister, networkPoliciesLister, namespaceLister, seedName),
+		recorder:            recorder,
+
+		endpointSynced:  endpointsInformer.Informer().HasSynced,
+		endpointsLister: endpointsLister,
+
+		networkPoliciesLister: networkPoliciesLister,
+		networkPoliciesSynced: networkPoliciesInformer.Informer().HasSynced,
+
+		namespaceSynced: namespaceInformer.Informer().HasSynced,
+		namespaceQueue:  namespaceQueue,
+		namespaceLister: namespaceLister,
+
+		shootNamespaceSelector: labels.SelectorFromSet(labels.Set{
+			v1beta1constants.DeprecatedGardenRole: v1beta1constants.GardenRoleShoot,
+			v1beta1constants.GardenRole:           v1beta1constants.GardenRoleShoot}),
+		gardenNamespaceSelector: labels.SelectorFromSet(labels.Set{
+			v1beta1constants.LabelRole: v1beta1constants.GardenNamespace}),
+
+		workerCh: make(chan int),
+	}
+
+	endpointsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			endpoints, ok := obj.(*corev1.Endpoints)
+			if !ok {
+				return false
+			}
+			return endpoints.Name == "kubernetes"
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc:    controller.endpointAdd,
+			UpdateFunc: controller.endpointUpdate,
+			DeleteFunc: controller.endpointDelete,
+		},
+	})
+
+	networkPoliciesInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			policy, ok := obj.(*networkingv1.NetworkPolicy)
+			if !ok {
+				return false
+			}
+			return policy.Name == helper.AllowToSeedAPIServer
+		},
+		Handler: cache.ResourceEventHandlerFuncs{
+			UpdateFunc: controller.networkPolicyUpdate,
+			DeleteFunc: controller.networkPolicyDelete,
+		},
+	})
+
+	namespaceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			ns, ok := obj.(*corev1.Namespace)
+			if !ok {
+				return false
+			}
+			namespaceLables := labels.Set(ns.Labels)
+			if controller.gardenNamespaceSelector.Matches(namespaceLables) {
+				return true
+			}
```

Comment:
I think we don't need this label selector as there will only be one `garden` namespace (https://github.com/gardener/gardener/pull/2339#discussion_r429064735).